server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -101,17 +107,44 @@ public ShardId getShardId() {
         throws IOException {
-        List<Exception> exceptionList = new ArrayList<>(transferSnapshot.getTranslogTransferMetadata().getCount());
-        Set<TransferFileSnapshot> toUpload = new HashSet<>(transferSnapshot.getTranslogTransferMetadata().getCount());
         long metadataBytesToUpload;
         long metadataUploadStartTime;
         long uploadStartTime;
         long prevUploadBytesSucceeded = remoteTranslogTransferTracker.getUploadBytesSucceeded();
         long prevUploadTimeInMillis = remoteTranslogTransferTracker.getTotalUploadTimeInMillis();
 
         try {
-            toUpload.addAll(fileTransferTracker.exclusionFilter(transferSnapshot.getTranslogFileSnapshots()));
-            toUpload.addAll(fileTransferTracker.exclusionFilter((transferSnapshot.getCheckpointFileSnapshots())));
+            boolean isObjectMetadataUploadSupportedInBlobStore = transferService.isObjectMetadataUploadSupported();
+
+            List<Exception> exceptionList = new ArrayList<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+            Set<TransferFileSnapshot> toUpload = new HashSet<>(transferSnapshot.getTranslogTransferMetadata().getCount());
+
+            // if the transferService of enabled blob store supports uploading object metadata, we don't need to transfer checkpoint file
+            // snapshots separately. We will transfer checkpoint data as metadata to translog file upload object.
+            if (isObjectMetadataUploadSupportedInBlobStore) {
+                Set<Tuple<TransferFileSnapshot, TransferFileSnapshot>> tlogAndCkpFilesTupleSet = transferSnapshot
+                    .getTranslogAndCheckpointFileSnapshotTupleSet();
+
+                for (Tuple<TransferFileSnapshot, TransferFileSnapshot> tuple : tlogAndCkpFilesTupleSet) {
+                    TransferFileSnapshot translogSnapshot = tuple.v1();
+                    TransferFileSnapshot checkpointSnapshot = tuple.v2();
+                    if (!fileTransferTracker.isAlreadyUploaded(translogSnapshot)) {
+                        Map<String, String> metadata = createCheckpointDataAsObjectMetadata(checkpointSnapshot);
+                        translogSnapshot.setTransferFileSnapshotMetadata(metadata);
+                        translogSnapshot.setTransferFileSnapshotMetadataFileName(checkpointSnapshot.getName());
+
+                        // close the checkpoint fileSnapshot now
+                        checkpointSnapshot.close();
+
+                        // add the translog fileSnapShot in toUpload set
+                        toUpload.add(translogSnapshot);
+                    }
+                }
+            } else {
+                toUpload.addAll(fileTransferTracker.exclusionFilter(transferSnapshot.getTranslogFileSnapshots()));
+                toUpload.addAll(fileTransferTracker.exclusionFilter(transferSnapshot.getCheckpointFileSnapshots()));
+            }
```

Comment:
Should we think of considering the translog and checkpoint as one generation and delegate the upload to the transferService. We will also need to update the `fileTransferTracker` accordingly to consider one generation as the tracking unit. 

The current code is appearing to complicate things and might make it a maintenance overhead in this state.

Reply:
I have a few point on this approach-

1. Delegating translog and checkpoint file to transferService and decide there to upload translog.tlog and translog.ckp file separatly or as single snapshot shouldn't be done there as this violate the principle of separation of concerns and transferService shouldn't have context of translog snapshot or checkpoint snapshot.
2. Changing fileTransferTracker for a generation unit instead of TransferFileSnapshot will also require us to maintain two fileTransferTracker as other remote stores should also work with the old approach.

However, we can further discuss on this offline!